src/auth/auth.ts

```diff
@@ -47,28 +56,58 @@ export class BBAuth {
     }
   } = {};
 
-  public static getUrl(
-    tokenizedUrl: string,
-    args?: BBAuthGetUrlArgs
-  ): Promise<string> {
-    // Returning a promise so eventually this could be enhanced
-    // to use a service discovery solution instead of using a convention.
-    const match = TOKENIZED_URL_REGEX.exec(tokenizedUrl);
-    let result = tokenizedUrl;
-    let zone = args ? args.zone : undefined;
+  public static extractUrl(
+    input: string,
+    zone?: string
+  ): OneBBUrlExtraction {
+    let url = input;
+    let scs;
+    let service;
+    let port;
+    let endpoint;
 
-    if (zone) {
-      zone = zone.replace('-', '');
-    }
+    const match = TOKENIZED_URL_REGEX.exec(input);
 
     if (match) {
+      scs = match[SCS_INDEX];
+      service = match[SERVICE_INDEX];
+      endpoint = match[ENDPOINT_INDEX];
+
       if (match[ZONE_INDEX]) {
-        zone = match[ZONE_INDEX].substring(1);
+        zone = match[ZONE_INDEX];
+      }
+      if (zone) {
+        zone = zone.replace('-', '');
       }
       // https://eng-pusa01.app.blackbaud.net/hub00/version
-      result = `https://${match[SCS_INDEX]}-${zone}.app.blackbaud.net/${match[SERVICE_INDEX]}/${match[ENDPOINT_INDEX]}`;
+      url = `https://${scs}-${zone}.app.blackbaud.net/${service}/${endpoint}`;
+
+      port = match[LOCAL_PORT_INDEX];
+      if (port) {
+        port = port.replace(':', '');
+      }
     }
-    return Promise.resolve(result);
+    return {
+      endpoint,
+      port,
+      scs,
+      service,
+      url,
+      zone
+    };
+  }
+
+  public static getUrl(
+    tokenizedUrl: string,
+    args?: BBAuthGetUrlArgs
+  ): Promise<string> {
```

Comment:
i like moving this promise bit to its own method. when the service discovery thing becomes a thing, in theory it's relegated to this method. 👏

Reply:
❤️